### PR TITLE
Prevent ~/.azkabanrc from becoming empty

### PR DESCRIPTION
**Pull request by kjmrknsn**
## Issue
`~/.azkabanrc` becomes empty when multiple processes/threads call AzkabanCLI APIs such as `azkaban.remote.Session.from_alias` simultaneously.

Example Code:
```
import azkaban.remote
import threading


def run():
    print('[thread %s] start' % threading.current_thread().name)
    azkaban.remote.Session.from_alias('dev')
    print('[thread %s] end' % threading.current_thread().name)


if __name__ == '__main__':
    print('start')
    threads = []
    for i in range(10):
        thread = threading.Thread(target=run, name="t" + str(i))
        thread.start()
        threads.append(thread)
    for thread in threads:
         thread.join()
    print('end')
```

Result:
```
$ python test.py
start
[thread t0] start
[thread t0] end
[thread t1] start
[thread t2] start
[thread t3] start
[thread t2] end
[thread t1] end
[thread t3] end
[thread t4] start
[thread t4] end
[thread t5] start
[thread t6] start
[thread t7] start
[thread t8] start
[thread t9] start
[thread t5] end
Exception in thread t6:
Traceback (most recent call last):
  File "/home/keiji.yoshida/local/python-2.7.14/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/home/keiji.yoshida/local/python-2.7.14/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "test.py", line 7, in run
    azkaban.remote.Session.from_alias('dev')
  File "/home/keiji.yoshida/local/python-2.7.14/lib/python2.7/site-packages/azkaban/remote.py", line 919, in from_alias
    raise AzkabanError('Alias not found: %r' % (alias, ))
AzkabanError: Alias not found: 'dev'
```

## Cause
https://github.com/mtth/azkaban/blob/master/azkaban/util.py#L78 reads an empty file when another process/thread calls https://github.com/mtth/azkaban/blob/master/azkaban/util.py#L89.

## Solution
The following 2 changes of `Config.__init__` reduce the possibility that this issue occurs.
* Call `self.save()` only if necessary.
* Make sure the length of `self.parser.sections()` is greater than 0 (i.e. The config instance read `~/.azkabanrc` which was not empty.) before `Config.__init__` ends.

### CODE REVIEW

**Comment by kjmrknsn** [comment]
@mtth My team is suffering from this issue. So, I'd be glad if you'd check and merge this PR, and bump the version of azkaban. Thanks.

File: azkaban/util.py [hunk 1]
```diff
@@ -69,20 +69,27 @@ class Config(object):
...
+  def __init__(self, path=None, max_read_cnt=10):
     self.parser = RawConfigParser()
     self.path = path or expanduser('~/.azkabanrc')
     # TODO: make the default path be configurable via an environment variable.
     if exists(self.path):
-      try:
-        self.parser.read(self.path)
-      except ParsingError:
-        raise AzkabanError('Invalid configuration file %r.', self.path)
+      read_cnt = 0
```
Comment (mtth) [inline azkaban/util.py]: IIUC the issue is with `save` calls happening concurrently with reads. If so, I think we can fix this by only saving when it is required. When the config file only has "new-style" aliases (which should be all the time, except for a one time conversion) no writing is required, so we can probably get away without any retry logic.

File: azkaban/util.py [hunk 2]
```diff
@@ -148,6 +155,9 @@ def _convert_aliases(self):
           parser.set(section, 'url', url)
           parser.set(section, 'verify', 'false') # Backwards compatibility.
       parser.remove_section('alias')
+      return True
+    else:
```
Comment (mtth) [inline azkaban/util.py]: Can you hoist the second branch earlier?

Something like:

```python
if not parser.has_section('alias'):
  return False # Nothing to do
for alias, url in parser.items('alias'):
  # ...
return True
```

File: azkaban/util.py (hunk 1 above)
Comment (kjmrknsn) [reply azkaban/util.py]: @mtth 
Thanks for reviewing this PR.

> IIUC the issue is with save calls happening concurrently with reads.

Yes, that's right. We can fix this issue without this retry logic. The only case a parser comes to read an empty file is that a parser's read process and save/write process of Azkaban session id happen simultaneously. I'm wondering we should implement this retry logic to handle this case or not.  I know implementing this kind of retry logic is a little bit messy and I'd like to know your idea.

File: azkaban/util.py (hunk 2 above)
Comment (kjmrknsn) [reply azkaban/util.py]: @mtth Thanks. I fixed this by https://github.com/mtth/azkaban/pull/28/commits/bc20fe9ad30c16e54697d0f11f0c90958f1cba50

**Comment by kjmrknsn** [comment]
@mtth Thanks for your reviewing. Eventually I has changed this PR only to fix `Config._convert_aliases`'s logic. Could you check this PR again? (Commits have been squashed.)

**Comment by mtth** [comment]
Thanks @kjmrknsn - released in `0.9.8`.

**Comment by kjmrknsn** [comment]
Thanks.

**Pull request merged**